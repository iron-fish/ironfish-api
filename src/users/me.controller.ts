/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Controller, Get, UseGuards } from '@nestjs/common';
import { ApiExcludeController } from '@nestjs/swagger';
import { MagicLinkGuard } from '../auth/guards/magic-link.guard';
import { Context } from '../common/decorators/context';
import { MagicLinkContext } from '../common/interfaces/magic-link-context';
import { SerializedUser } from './interfaces/serialized-user';
import { serializedUserFromRecord } from './utils/user-translator';

@ApiExcludeController()
@Controller('me')
export class MeController {
  @Get()
  @UseGuards(MagicLinkGuard)
  me(@Context() { user }: MagicLinkContext): SerializedUser {
    return serializedUserFromRecord(user);
  }
}
