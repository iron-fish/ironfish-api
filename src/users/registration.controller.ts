/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Controller, Get, Param, Res } from '@nestjs/common';
import { ApiExcludeEndpoint } from '@nestjs/swagger';
import { Response } from 'express';
import { ApiConfigService } from '../api-config/api-config.service';
import { UsersService } from './users.service';

@Controller('registration')
export class RegistrationController {
  constructor(
    private readonly config: ApiConfigService,
    private readonly usersService: UsersService,
  ) {}

  @ApiExcludeEndpoint()
  @Get(':token/confirm')
  async confirm(
    @Param('token') token: string,
    @Res() res: Response,
  ): Promise<void> {
    const user = await this.usersService.findByConfirmationToken(token);
    if (!user || user.confirmed_at) {
      const message = 'Invalid confirmation token';
      res.redirect(
        `${this.config.get<string>(
          'INCENTIVIZED_TESTNET_URL',
        )}/login?toast=${Buffer.from(message).toString('base64')}`,
      );
    } else {
      await this.usersService.confirm(user);
      res.redirect(
        `${this.config.get<string>(
          'INCENTIVIZED_TESTNET_URL',
        )}/login?confirmed=true`,
      );
    }
  }
}
