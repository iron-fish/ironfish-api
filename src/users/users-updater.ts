/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable, UnprocessableEntityException } from '@nestjs/common';
import { BlocksService } from '../blocks/blocks.service';
import { PrismaService } from '../prisma/prisma.service';
import { UpdateUserOptions } from './interfaces/update-user-options';
import { UsersService } from './users.service';
import { User } from '.prisma/client';

@Injectable()
export class UsersUpdater {
  constructor(
    private readonly blocksService: BlocksService,
    private readonly prisma: PrismaService,
    private readonly usersService: UsersService,
  ) {}

  async update(user: User, options: UpdateUserOptions): Promise<User> {
    return this.prisma.$transaction(async (prisma) => {
      const { discord, graffiti, telegram } = options;

      if (graffiti && user.graffiti !== graffiti) {
        const minedBlocksForCurrentGraffiti =
          await this.blocksService.countByGraffiti(user.graffiti, prisma);
        if (minedBlocksForCurrentGraffiti > 0) {
          throw new UnprocessableEntityException({
            code: 'user_graffiti_already_used',
            message: `Current graffiti '${user.graffiti}' has already mined blocks`,
          });
        }

        const minedBlocksForNewGraffiti =
          await this.blocksService.countByGraffiti(graffiti, prisma);
        if (minedBlocksForNewGraffiti > 0) {
          throw new UnprocessableEntityException({
            code: 'duplicate_block_graffiti',
            message: `Blocks with '${graffiti}' already exist`,
          });
        }
      }

      const users = await this.usersService.findDuplicateUser(
        user,
        options,
        prisma,
      );
      if (users.length) {
        const duplicateUser = users[0];
        let error;

        if (discord && discord === duplicateUser.discord) {
          error = {
            code: 'duplicate_user_discord',
            message: `User with Discord '${discord}' already exists`,
          };
        } else if (graffiti && graffiti === duplicateUser.graffiti) {
          error = {
            code: 'duplicate_user_graffiti',
            message: `User with graffiti '${graffiti}' already exists`,
          };
        } else if (telegram && telegram === duplicateUser.telegram) {
          error = {
            code: 'duplicate_user_telegram',
            message: `User with Telegram '${telegram}' already exists`,
          };
        } else {
          throw new Error('Unexpected database response');
        }

        throw new UnprocessableEntityException(error);
      }

      return this.usersService.update(user, options, prisma);
    });
  }
}
