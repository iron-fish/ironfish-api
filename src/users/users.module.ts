/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { EventsModule } from '../events/events.module';
import { PrismaModule } from '../prisma/prisma.module';
import { UsersService } from './users.service';

@Module({
  exports: [UsersService],
  imports: [EventsModule, PrismaModule],
  providers: [UsersService],
})
export class UsersModule {}
