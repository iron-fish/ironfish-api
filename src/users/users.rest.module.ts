/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { EventsModule } from '../events/events.module';
import { MeController } from './me.controller';
import { UsersController } from './users.controller';
import { UsersModule } from './users.module';

@Module({
  controllers: [MeController, UsersController],
  imports: [EventsModule, UsersModule],
})
export class UsersRestModule {}
