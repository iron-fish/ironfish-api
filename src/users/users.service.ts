/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import {
  Injectable,
  NotFoundException,
  UnprocessableEntityException,
} from '@nestjs/common';
import is from '@sindresorhus/is';
import { ulid } from 'ulid';
import { ApiConfigService } from '../api-config/api-config.service';
import { DEFAULT_LIMIT, MAX_LIMIT } from '../common/constants';
import { SortOrder } from '../common/enums/sort-order';
import { standardizeEmail } from '../common/utils/email';
import { PostmarkService } from '../postmark/postmark.service';
import { PrismaService } from '../prisma/prisma.service';
import { BasePrismaClient } from '../prisma/types/base-prisma-client';
import { CreateUserDto } from './dto/create-user.dto';
import { ListUsersWithRankOptions } from './interfaces/list-by-rank-options';
import { ListUsersOptions } from './interfaces/list-users-options';
import { SerializedUserWithRank } from './interfaces/serialized-user-with-rank';
import { UpdateUserOptions } from './interfaces/update-user-options';
import { Prisma, User } from '.prisma/client';

@Injectable()
export class UsersService {
  constructor(
    private readonly config: ApiConfigService,
    private readonly postmarkService: PostmarkService,
    private readonly prisma: PrismaService,
  ) {}

  async findConfirmed(id: number): Promise<User | null> {
    const record = await this.prisma.user.findUnique({
      where: { id },
    });
    if (record === null || record.confirmed_at === null) {
      return null;
    }
    return record;
  }

  async findOrThrow(id: number): Promise<User> {
    const record = await this.findConfirmed(id);
    if (record === null) {
      throw new NotFoundException();
    }
    return record;
  }

  async findByGraffiti(
    graffiti: string,
    prisma?: BasePrismaClient,
  ): Promise<User | null> {
    const client = prisma ?? this.prisma;
    return client.user.findFirst({
      where: {
        graffiti,
      },
    });
  }

  async findByGraffitiOrThrow(graffiti: string): Promise<User> {
    const record = await this.findByGraffiti(graffiti);
    if (!record) {
      throw new NotFoundException();
    }
    return record;
  }

  async findByEmailOrThrow(email: string): Promise<User> {
    const record = await this.findByEmail(email);
    if (!record) {
      throw new NotFoundException();
    }
    return record;
  }

  async findByEmail(email: string): Promise<User | null> {
    return this.prisma.user.findFirst({
      where: {
        email,
      },
    });
  }

  async listByEmail(email: string): Promise<User[]> {
    email = standardizeEmail(email);
    return this.prisma.user.findMany({
      where: {
        email,
      },
    });
  }

  async findByConfirmationToken(token: string): Promise<User | null> {
    return this.prisma.user.findUnique({
      where: {
        confirmation_token: token,
      },
    });
  }

  async create({
    email,
    graffiti,
    country_code: countryCode,
    discord,
    telegram,
    github,
  }: CreateUserDto): Promise<User> {
    email = standardizeEmail(email);
    const existingRecord = await this.prisma.user.findFirst({
      where: {
        OR: [
          {
            email,
          },
          {
            graffiti,
          },
          {
            github,
          },
        ],
      },
    });
    if (existingRecord) {
      throw new UnprocessableEntityException(
        `User already exists for '${
          graffiti === existingRecord.graffiti ? graffiti : email
        }'`,
      );
    }
    const [user] = await this.prisma.$transaction([
      this.prisma.user.create({
        data: {
          email,
          graffiti,
          discord,
          telegram,
          github,
          country_code: countryCode,
          confirmation_token: ulid(),
        },
      }),
    ]);
    await this.postmarkService.send({
      alias: 'incentivized-testnet-confirmation',
      templateModel: {
        action_url: `${this.config.get<string>('API_URL')}/registration/${
          user.confirmation_token
        }/confirm`,
        graffiti: user.graffiti,
      },
      to: user.email,
    });
    return user;
  }

  async list(options: ListUsersOptions): Promise<{
    data: User[];
    hasNext: boolean;
    hasPrevious: boolean;
  }> {
    const cursorId = options.before ?? options.after;
    const cursor = cursorId ? { id: cursorId } : undefined;
    const direction = options.before !== undefined ? -1 : 1;
    const limit =
      direction * Math.min(MAX_LIMIT, options.limit || DEFAULT_LIMIT);
    const orderBy = { id: SortOrder.DESC };
    const skip = cursor ? 1 : 0;
    const where = {
      country_code: options.countryCode,
      graffiti: {
        contains: options.search,
      },
      confirmed_at: {
        not: null,
      },
    };
    const data = await this.prisma.user.findMany({
      cursor,
      orderBy,
      skip,
      take: limit,
      where,
    });
    return {
      data,
      ...(await this.getListMetadata(data, where, orderBy)),
    };
  }

  private async getListMetadata(
    data: User[],
    where: Prisma.UserWhereInput,
    orderBy: Prisma.Enumerable<Prisma.UserOrderByWithRelationInput>,
  ): Promise<{ hasNext: boolean; hasPrevious: boolean }> {
    const { length } = data;
    if (length === 0) {
      return {
        hasNext: false,
        hasPrevious: false,
      };
    }
    const nextRecords = await this.prisma.user.findMany({
      where,
      orderBy,
      cursor: { id: data[length - 1].id },
      skip: 1,
      take: 1,
    });
    const previousRecords = await this.prisma.user.findMany({
      where,
      orderBy,
      cursor: { id: data[0].id },
      skip: 1,
      take: -1,
    });
    return {
      hasNext: nextRecords.length > 0,
      hasPrevious: previousRecords.length > 0,
    };
  }

  async listWithRank({
    after,
    before,
    limit,
    search,
    countryCode,
    eventType,
  }: ListUsersWithRankOptions): Promise<{
    data: SerializedUserWithRank[];
    hasNext: boolean;
    hasPrevious: boolean;
  }> {
    let rankCursor: number;
    const cursorId = before ?? after;
    if (cursorId !== undefined) {
      rankCursor = await this.getRank(cursorId);
    } else {
      // Ranks start at 1, so get everything after 0
      rankCursor = 0;
    }
    const searchFilter = `%${search ?? ''}%`;
    const query = `
      SELECT
        id,
        graffiti,
        total_points,
        country_code,
        last_login_at,
        rank
      FROM
        (
          SELECT
            id,
            graffiti,
            COALESCE(user_latest_events.total_points, 0) as total_points,
            country_code,
            last_login_at,
            RANK () OVER ( 
              ORDER BY 
                COALESCE(user_latest_events.total_points, 0) DESC,
                COALESCE(latest_event_occurred_at, NOW()) ASC,
                created_at ASC
            ) AS rank 
          FROM
            users
          LEFT JOIN
            (
              SELECT
                user_id,
                SUM(points) as total_points,
                MAX(occurred_at) AS latest_event_occurred_at
              FROM
                (
                  SELECT
                    user_id,
                    occurred_at,
                    points
                  FROM
                    events
                  WHERE
                    deleted_at IS NULL AND
                    CASE WHEN $6::event_type IS NOT NULL
                      THEN
                        type = $6
                      ELSE
                        TRUE
                    END
                ) filtered_events
              GROUP BY
                user_id
            ) user_latest_events
          ON
            user_latest_events.user_id = users.id
          WHERE
            confirmed_at IS NOT NULL
        ) user_ranks
      WHERE
        graffiti ILIKE $1 AND
        CASE WHEN $2
          THEN
            rank > $3
          ELSE
            rank < $3
        END AND
        CASE WHEN $5::text IS NOT NULL
          THEN
            country_code = $5
          ELSE
            TRUE
        END
      ORDER BY
        rank ASC
      LIMIT
        $4`;

    const data = await this.prisma.$queryRawUnsafe<SerializedUserWithRank[]>(
      query,
      searchFilter,
      before === undefined,
      rankCursor,
      limit,
      countryCode,
      eventType,
    );
    return {
      data,
      ...(await this.getListWithRankMetadata(
        data,
        query,
        searchFilter,
        countryCode,
        eventType,
      )),
    };
  }

  private async getListWithRankMetadata(
    data: SerializedUserWithRank[],
    query: string,
    searchFilter: string,
    countryCode?: string,
    eventType?: string,
  ): Promise<{ hasNext: boolean; hasPrevious: boolean }> {
    const { length } = data;
    if (length === 0) {
      return {
        hasNext: false,
        hasPrevious: false,
      };
    }
    const nextRecords = await this.prisma.$queryRawUnsafe<
      SerializedUserWithRank[]
    >(
      query,
      searchFilter,
      true,
      data[length - 1].rank,
      1,
      countryCode,
      eventType,
    );
    const previousRecords = await this.prisma.$queryRawUnsafe<
      SerializedUserWithRank[]
    >(query, searchFilter, false, data[0].rank, 1, countryCode, eventType);
    return {
      hasNext: nextRecords.length > 0,
      hasPrevious: previousRecords.length > 0,
    };
  }

  async updateLastLoginAt(user: User): Promise<User> {
    return this.prisma.$transaction(async (prisma) => {
      return prisma.user.update({
        where: {
          id: user.id,
        },
        data: {
          last_login_at: new Date().toISOString(),
        },
      });
    });
  }

  async getRank(userOrId: User | number): Promise<number> {
    let id: number;
    if (typeof userOrId === 'number') {
      const record = await this.findOrThrow(userOrId);
      id = record.id;
    } else {
      id = userOrId.id;
    }
    const rankResponse = await this.prisma.$queryRawUnsafe<{ rank: number }[]>(
      `SELECT
        id,
        rank
      FROM
        (
          SELECT
            id,
            RANK () OVER ( 
              ORDER BY 
                total_points DESC,
                COALESCE(latest_event_occurred_at, NOW()) ASC,
                created_at ASC
            ) AS rank 
          FROM
            users
          LEFT JOIN
            (
              SELECT
                user_id,
                MAX(occurred_at) AS latest_event_occurred_at
              FROM
                (
                  SELECT
                    user_id,
                    occurred_at
                  FROM
                    events
                  WHERE
                    deleted_at IS NULL
                ) filtered_events
              GROUP BY
                user_id
            ) user_latest_events
          ON
            user_latest_events.user_id = users.id
          WHERE
            users.confirmed_at IS NOT NULL
        ) user_ranks
      WHERE
        id = $1`,
      id,
    );
    if (
      !is.array(rankResponse) ||
      rankResponse.length !== 1 ||
      !is.object(rankResponse[0]) ||
      !('id' in rankResponse[0]) ||
      !('rank' in rankResponse[0])
    ) {
      throw new Error('Unexpected database response');
    }
    return rankResponse[0].rank;
  }

  async confirm(user: User): Promise<User> {
    return this.prisma.$transaction(async (prisma) => {
      return prisma.user.update({
        where: {
          id: user.id,
        },
        data: {
          confirmed_at: new Date().toISOString(),
        },
      });
    });
  }

  async findDuplicateUser(
    user: User,
    options: UpdateUserOptions,
    client: BasePrismaClient,
  ): Promise<User[]> {
    const { discord, graffiti, telegram } = options;
    return client.user.findMany({
      where: {
        confirmed_at: {
          not: null,
        },
        OR: [{ discord }, { graffiti }, { telegram }],
        NOT: {
          id: user.id,
        },
      },
    });
  }

  async update(
    user: User,
    options: UpdateUserOptions,
    client: BasePrismaClient,
  ): Promise<User> {
    const { countryCode, discord, graffiti, telegram } = options;
    return client.user.update({
      data: {
        country_code: countryCode,
        discord,
        graffiti,
        telegram,
      },
      where: {
        id: user.id,
      },
    });
  }
}
