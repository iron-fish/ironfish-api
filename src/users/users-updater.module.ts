/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { BlocksModule } from '../blocks/blocks.module';
import { PrismaModule } from '../prisma/prisma.module';
import { UsersModule } from './users.module';
import { UsersUpdater } from './users-updater';

@Module({
  exports: [UsersUpdater],
  imports: [BlocksModule, PrismaModule, UsersModule],
  providers: [UsersUpdater],
})
export class UsersUpdaterModule {}
