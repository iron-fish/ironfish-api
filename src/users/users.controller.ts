/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import {
  Body,
  Controller,
  ForbiddenException,
  Get,
  HttpStatus,
  Param,
  ParseIntPipe,
  Post,
  Put,
  Query,
  UnprocessableEntityException,
  UseGuards,
  ValidationPipe,
} from '@nestjs/common';
import {
  ApiExcludeEndpoint,
  ApiOperation,
  ApiParam,
  ApiTags,
} from '@nestjs/swagger';
import { MagicLinkGuard } from '../auth/guards/magic-link.guard';
import { DEFAULT_LIMIT, MAX_LIMIT, MS_PER_DAY } from '../common/constants';
import { Context } from '../common/decorators/context';
import { MetricsGranularity } from '../common/enums/metrics-granularity';
import { MagicLinkContext } from '../common/interfaces/magic-link-context';
import { PaginatedList } from '../common/interfaces/paginated-list';
import { EventsService } from '../events/events.service';
import { SerializedEventMetrics } from '../events/interfaces/serialized-event-metrics';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { UserMetricsQueryDto } from './dto/user-metrics-query.dto';
import { UsersQueryDto } from './dto/users-query.dto';
import { SerializedUser } from './interfaces/serialized-user';
import { SerializedUserMetrics } from './interfaces/serialized-user-metrics';
import { SerializedUserWithRank } from './interfaces/serialized-user-with-rank';
import { UsersService } from './users.service';
import { UsersUpdater } from './users-updater';
import {
  serializedUserFromRecord,
  serializedUserFromRecordWithRank,
} from './utils/user-translator';
import { EventType, User } from '.prisma/client';

const MAX_SUPPORTED_TIME_RANGE_IN_DAYS = 30;

@ApiTags('Users')
@Controller('users')
export class UsersController {
  constructor(
    private readonly eventsService: EventsService,
    private readonly usersService: UsersService,
    private readonly usersUpdater: UsersUpdater,
  ) {}

  @ApiOperation({ summary: 'Gets a specific User' })
  @ApiParam({ description: 'Unique User identifier', name: 'id' })
  @Get(':id')
  async find(
    @Param(
      'id',
      new ParseIntPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
      }),
    )
    id: number,
  ): Promise<SerializedUser> {
    const user = await this.usersService.findOrThrow(id);
    return serializedUserFromRecordWithRank(
      user,
      await this.usersService.getRank(user),
    );
  }

  @ApiOperation({ summary: 'Gets metrics for a specific User' })
  @ApiParam({ description: 'Unique User identifier', name: 'id' })
  @Get(':id/metrics')
  async metrics(
    @Param(
      'id',
      new ParseIntPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
      }),
    )
    id: number,
    @Query(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    query: UserMetricsQueryDto,
  ): Promise<SerializedUserMetrics> {
    const { isValid, error } = this.isValidMetricsQuery(query);
    if (!isValid) {
      throw new UnprocessableEntityException(error);
    }

    let eventMetrics: Record<EventType, SerializedEventMetrics>;
    let points: number;
    const { start, end, granularity } = query;
    if (granularity === MetricsGranularity.LIFETIME) {
      const user = await this.usersService.findOrThrow(id);
      eventMetrics = await this.eventsService.getLifetimeEventMetricsForUser(
        user,
      );
      points = user.total_points;
    } else {
      if (start === undefined || end === undefined) {
        throw new UnprocessableEntityException(
          'Must provide time range for "TOTAL" requests',
        );
      }
      const user = await this.usersService.findOrThrow(id);
      ({ eventMetrics, points } =
        await this.eventsService.getTotalEventMetricsAndPointsForUser(
          user,
          start,
          end,
        ));
    }

    return {
      user_id: id,
      granularity,
      points,
      metrics: {
        blocks_mined: eventMetrics[EventType.BLOCK_MINED],
        bugs_caught: eventMetrics[EventType.BUG_CAUGHT],
        community_contributions: eventMetrics[EventType.COMMUNITY_CONTRIBUTION],
        pull_requests_merged: eventMetrics[EventType.PULL_REQUEST_MERGED],
        social_media_contributions:
          eventMetrics[EventType.SOCIAL_MEDIA_PROMOTION],
      },
    };
  }

  private isValidMetricsQuery({
    start,
    end,
    granularity,
  }: UserMetricsQueryDto): {
    isValid: boolean;
    error?: string;
  } {
    if (
      granularity !== MetricsGranularity.LIFETIME &&
      granularity !== MetricsGranularity.TOTAL
    ) {
      return {
        isValid: false,
        error: '"granularity" must be "lifetime" or "total"',
      };
    }

    if (start !== undefined && end !== undefined) {
      if (granularity === MetricsGranularity.LIFETIME) {
        return {
          isValid: false,
          error: 'Cannot provide time range for "LIFETIME" requests',
        };
      }
      if (start >= end) {
        return {
          isValid: false,
          error: '"start" must be stricly less than "end"',
        };
      }

      const diffInMs = end.getTime() - start.getTime();
      const diffInDays = diffInMs / MS_PER_DAY;
      if (diffInDays > MAX_SUPPORTED_TIME_RANGE_IN_DAYS) {
        return {
          isValid: false,
          error: 'Time range too long',
        };
      }
    }
    return { isValid: true };
  }

  @ApiOperation({ summary: 'Returns a paginated list of users' })
  @Get()
  async list(
    @Query(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    {
      after,
      before,
      limit,
      order_by: orderBy,
      country_code: countryCode,
      event_type: eventType,
      search,
    }: UsersQueryDto,
  ): Promise<PaginatedList<SerializedUser | SerializedUserWithRank>> {
    if (orderBy !== undefined) {
      const { data, hasNext, hasPrevious } =
        await this.usersService.listWithRank({
          after,
          before,
          limit: Math.min(MAX_LIMIT, limit || DEFAULT_LIMIT),
          search,
          countryCode,
          eventType,
        });
      return {
        object: 'list',
        data,
        metadata: {
          has_next: hasNext,
          has_previous: hasPrevious,
        },
      };
    }
    const { data, hasNext, hasPrevious } = await this.usersService.list({
      after,
      before,
      limit,
      search,
      countryCode,
      eventType,
    });
    return {
      object: 'list',
      data: data.map((user) => serializedUserFromRecord(user)),
      metadata: {
        has_next: hasNext,
        has_previous: hasPrevious,
      },
    };
  }

  @ApiExcludeEndpoint()
  @Post()
  async create(
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    dto: CreateUserDto,
  ): Promise<User> {
    return this.usersService.create(dto);
  }

  @ApiExcludeEndpoint()
  @Put(':id')
  @UseGuards(MagicLinkGuard)
  async update(
    @Context() { user }: MagicLinkContext,
    @Param(
      'id',
      new ParseIntPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
      }),
    )
    id: number,
    @Body(
      new ValidationPipe({
        errorHttpStatusCode: HttpStatus.UNPROCESSABLE_ENTITY,
        transform: true,
      }),
    )
    dto: UpdateUserDto,
  ): Promise<User> {
    if (id !== user.id) {
      throw new ForbiddenException();
    }
    return this.usersUpdater.update(user, {
      countryCode: dto.country_code,
      discord: dto.discord,
      graffiti: dto.graffiti,
      telegram: dto.telegram,
    });
  }
}
