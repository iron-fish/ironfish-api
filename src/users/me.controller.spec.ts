/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { HttpStatus, INestApplication } from '@nestjs/common';
import faker from 'faker';
import request from 'supertest';
import { v4 as uuid } from 'uuid';
import { MagicLinkService } from '../magic-link/magic-link.service';
import { bootstrapTestApp } from '../test/test-app';
import { UsersService } from './users.service';

describe('MeController', () => {
  let app: INestApplication;
  let magicLinkService: MagicLinkService;
  let usersService: UsersService;

  beforeAll(async () => {
    app = await bootstrapTestApp();
    magicLinkService = app.get(MagicLinkService);
    usersService = app.get(UsersService);
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('GET /me', () => {
    describe('with no logged in user', () => {
      it('returns a 401', async () => {
        const { body } = await request(app.getHttpServer())
          .get('/me')
          .expect(HttpStatus.UNAUTHORIZED);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a logged in user', () => {
      it('returns the user', async () => {
        const user = await usersService.create({
          email: faker.internet.email(),
          graffiti: uuid(),
          country_code: faker.address.countryCode('alpha-3'),
        });
        await usersService.confirm(user);

        jest
          .spyOn(magicLinkService, 'getEmailFromHeader')
          .mockImplementationOnce(() => Promise.resolve(user.email));

        const { body } = await request(app.getHttpServer())
          .get('/me')
          .set('Authorization', 'did-token')
          .expect(HttpStatus.OK);

        expect(body).toMatchObject({
          id: user.id,
          country_code: user.country_code,
          graffiti: user.graffiti,
        });
      });
    });
  });
});
