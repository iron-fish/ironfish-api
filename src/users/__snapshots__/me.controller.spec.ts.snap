// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeController GET /me with no logged in user returns a 401 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;
